### DISCUSSION TITLE
Please help with FileLoop

### DISCUSSION THREAD

**Discussion by BODYPRINT** [pschatzmann/arduino-audio-tools#1844] [Q&A] [answered]
Hi, apologise because I'm a bit of a C newbie.
I'm making an ESP32 soundboard for my grandson and can load and play wav samples from the SD Card, but I'm trying to get them to loop, and I can't seem to get any of the FileLoop class to work.

Tried doing, position and seek in the main loop but I think the buffer makes this not work correctly.

Here's my current code, can I please get some help implementing FileLoop thanks?
Any other tips on how to make the code better woiuld be appreciated as well. :)

Thanks in advance.

```
#include <SD.h>
#include <SPI.h>
#include "AudioTools.h"

#define BUTTON_PIN 15

const int chipSelect = 5;
AudioInfo info(16000, 1, 16);
PWMAudioOutput out;
WAVDecoder wav;
EncodedAudioOutput decoder(&out, &wav);  // Decoding stream
StreamCopy copier;
File audioFile;
const char engine[] = "/start.wav";
File audioStart;

void setup() {
  pinMode(BUTTON_PIN, INPUT_PULLUP);
  Serial.begin(9600);
  AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);

  // setup file
  Serial.println("Opening Sound File beep16.wav");
  SD.begin(chipSelect);
  audioFile = SD.open("/beep16.wav");
  if (audioFile == false) {
    Serial.println("Failed to open file");
  }

  // start Output
  auto cfgi = out.defaultConfig(TX_MODE);
  cfgi.copyFrom(info);
  wav.setAudioInfo(info);
  wav.begin();
  out.begin(cfgi);

  // setup decoder
  decoder.begin(info);

  // begin copy
  copier.begin(decoder, audioFile);
  AudioToolsLogger.setLogLevel(AudioToolsLogLevel::Warning);
}

void loop() {
  int buttonState = digitalRead(BUTTON_PIN);
  if (!buttonState) {
    copier.copy();
    if (audioFile.position()>=7886){
      audioFile.seek(4536);
    }
  }
}
```

**Comment by pschatzmann** [comment, marked as answer]
I guess your issue is in the loop logic:
- when the file is < 7886, there will be never a seek (): I would check the result of copy(): if it is 0 you are at the end of the file
- maybe you should debounce the read or just handle state changes ? - I don't quite understand your setup and intention here
- If you want to loop over the complete file you can use the FileLoop class
- Your file positions are rather small so you just play a couple of milliseconds of audio

Just keep the log level info in the loop and add some log statement in your loop, so that you see when your seek is called.
